test/rules/require_trailing_comma.dart

```diff
@@ -0,0 +1,223 @@
+      },
+      param3: 'test',
+    );
+
+    test('test', 'test', param3: () {
+      // Function literal implemented using curly braces.
+    }); // LINT
+
+    test(
+      'test',
+      'test',
+      param3: () {
+        // Function literal implemented using curly braces.
+      },
+    );
+
+    test(
+          () {
+        // Function literal implemented using curly braces.
+      },
+      'test',
+    );
+
+    test(() {
+      // Function literal implemented using curly braces.
+    }, 'test'); // LINT
+
+    test('map literal', {
+      'one': 'test',
+      'two': 'test',
+    });
+
+    test({
+      'one': 'test',
+      'two': 'test',
+    }, 'map literal'); // LINT
+
+    test('set literal', {
+    'one',
+    'two',
+    });
+
+    test({
+    'one',
+    'two',
+    }, 'set literal'); // LINT
+
+    test('list literal', [
+      'one',
+      'two',
+    ]);
+
+    test([
+      'one',
+      'two',
+    ], 'list literal'); // LINT
+
+    test(
+      'no exception for set literal as it fits entirely on 1 line',
+      const {'one', 'two', 'three'},
+    );
+
+    test('no exception for set literal as it fits entirely on 1 line',
+        const {'one', 'two', 'three'}); // LINT
+
+    test('exception for set literal as it spans multiple lines', const {
+    'one',
+    'two',
+    'three',
+    });
+
+    test('exception for set literal as it spans multiple lines', const <
+        AnExtremelyLongClassNameOneTwoThreeFourFiveSixSevenEightNineTen>{});
+
+    test(
+      'no exception for array literal as it fits entirely on 1 line',
+      const ['one', 'two', 'three'],
+    );
+
+    test('no exception for array literal as it fits entirely on 1 line',
+        const ['one', 'two', 'three']); // LINT
+
+    test('exception for array literal as it spans multiple lines', const [
+      'one',
+      'two',
+      'three',
+    ]);
+
+    test('exception for array literal as it spans multiple lines', const <
+        AnExtremelyLongClassNameOneTwoThreeFourFiveSixSevenEightNineTen>[]);
+
+    test(
+      'no exception for map literal as it fits entirely on 1 line',
+      const {'one': '1', 'two': '2', 'three': '3'},
+    );
+
+    test('no exception for map literal as it fits entirely on 1 line',
+        const {'one': '1', 'two': '2', 'three': '3'}); // LINT
+
+    test('exception for map literal as it spans multiple lines', const {
+      'one': '1',
+      'two': '2',
+      'three': '3',
+    });
+
+    test('exception for map literal as it spans multiple lines', const <String,
+        AnExtremelyLongClassNameOneTwoThreeFourFiveSixSevenEightNineTen>{});
+
+    test(
+      'no exception for function literal as it fits entirely on 1 line',
+          () {},
+    );
+
+    test('no exception for function literal as it fits entirely on 1 line',
+            () {}); // LINT
+
+    assert(true);
+
+    assert('a very very very very very very very very very long string'
+        .isNotEmpty); // LINT
```

Comment:
Heads up: I suspect this part will be controversial. I don't have a strong opinion yet, but I suspect for many people, requiring a trailing comma for a single parameter/argument (on a long line) may be a bridge too far.